addons/api/addon/workers/shared-service.js

```diff
@@ -0,0 +1,242 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import { v4 as uuidv4 } from 'uuid';
+
+let clientChannel = null;
+const requestsInFlight = new Map();
+
+export default (serviceName, target, onProviderChange) => {
+  const commonChannel = new BroadcastChannel(
+    `shared-service-common-channel-${serviceName}`,
+  );
+
+  let readyResolve = null;
+  const status = {
+    // Set a promise that doesn't resolve until our setup is finished
+    ready: new Promise((resolve) => {
+      readyResolve = resolve;
+    }),
+    // Check if there's an existing provider
+    isServiceProvider: (async () => {
+      const locks = await navigator.locks.query();
+      const isProvider = !locks.held?.find((lock) => lock.name === serviceName);
+      if (!isProvider) {
+        await onNotProvider();
+      }
+      return isProvider;
+    })(),
+  };
+
+  // This lock is used to ensure that only one tab can become the provider with an indefinite lock.
+  navigator.locks.request(serviceName, { mode: 'exclusive' }, async () => {
+    await onBecomeProvider();
+    await new Promise(() => {});
+  });
+
+  const onNotProvider = async () => {
+    const clientId = getClientId();
+    clientChannel = new BroadcastChannel(
+      getClientBroadcastChannelName(clientId, serviceName),
+    );
+
+    const register = async () =>
+      new Promise((resolve) => {
+        const onRegisteredListener = ({ data }) => {
+          if (data.clientId === clientId && data.type === 'registered') {
+            commonChannel.removeEventListener('message', onRegisteredListener);
+            resolve();
+          }
+        };
+
+        // Add an event listener to check for the registration response
+        commonChannel.addEventListener('message', onRegisteredListener);
+        commonChannel.postMessage({ type: 'register', clientId });
+      });
+
+    commonChannel.addEventListener('message', async ({ data }) => {
+      // Check if this was a provider change. If so and we're not the new provider, re-register.
+      if (data.type === 'providerChange' && !(await status.isServiceProvider)) {
+        await onProviderChange?.(await status.isServiceProvider);
+        await register();
+
+        // If there were requests in flight when the provider was changed,
+        // we need to resend them to the new provider.
+        if (requestsInFlight.size > 0) {
```

Comment:
nice resend logic here! 

do we need to think about debouncing or adding any guard here in case of multiple provider changes? Might not be an issue right now, but I was just wondering if this is something to consider.

Reply:
I'm not sure if we need to worry about that, it's already likely uncommon for requests to be in flight that will need to be re-tried, and even if there are multiple provider changes happening around the same time, what ends up happening is just some wasted requests as they don't finish.